src/System.Management.Automation/engine/serialization.cs

```diff
@@ -3986,19 +3987,104 @@ private object ReadListContainer(ContainerType ct)
             return list;
         }
 
+        /// <summary>
+        /// Some ordered or non-ordered Dictionaty methods.
+        /// </summary>
+        private class PSDictionary
+        {
+            private Hashtable table;
+            private OrderedDictionary orderedDictionary;
+            private readonly bool _isOrdered;
+            private int _keyClashFoundIteration;
+            private bool _containsKey;
+
+            public PSDictionary(bool isOrdered) {
+                _isOrdered = isOrdered;
+                _keyClashFoundIteration = 0;
+
+                if (_isOrdered) {
+                    orderedDictionary = new OrderedDictionary(StringComparer.CurrentCultureIgnoreCase);
+                } else {
+                    table = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+                }
+            }
+
+            public object DictionaryObject { get { return _isOrdered ? orderedDictionary : table; } }
+
+            public void Add(object key, object value) {
+                // On the first collision, copy the hash table to one that uses the default comparer.
+                _containsKey = _isOrdered ? orderedDictionary.Contains(key) : table.ContainsKey(key);
+                if (_keyClashFoundIteration == 0 && _containsKey)
```

Comment:
Can you add a test for a key clash?

Reply:
Both kinds

Reply:
Added test cases.  See the general discussion comments.